File: modules/runtime/libs/golang/stub.go
package iot

import (
	"os"
	"sync"

	log "github.com/sirupsen/logrus"
	"google.golang.org/grpc"

	rpc "github.com/slamy-solutions/openbp/modules/runtime/libs/golang/manager/rpc"
	runtime "github.com/slamy-solutions/openbp/modules/runtime/libs/golang/manager/runtime"
)

func getConfigEnv(key string, fallback string) string {
	if value, ok := os.LookupEnv(key); ok {
		return value
	}
	return fallback
}

type ManagerService struct {
	Runtime runtime.RuntimeServiceClient
	RPC     rpc.RPCServiceClient
}

type GrpcServiceConfig struct {
	enabled bool
	url     string
}

type StubConfig struct {
	logger *log.Logger

	manager GrpcServiceConfig
}

func NewStubConfig() *StubConfig {
	return &StubConfig{
		logger: log.StandardLogger(),
		manager: GrpcServiceConfig{
			enabled: false,
			url:     "",
		},
	}
}

func (sc *StubConfig) WithLogger(logger *log.Logger) *StubConfig {
	sc.logger = logger
	return sc
}

func (sc *StubConfig) WithManagerService(conf ...GrpcServiceConfig) *StubConfig {
	if len(conf) != 0 {
		sc.manager = conf[0]
	} else {
		sc.manager = GrpcServiceConfig{
			enabled: true,
			url:     getConfigEnv("RUNTIME_MANAGER_URL", "runtime_manager:80"),
		}
	}
	return sc
}

type RuntimeStub struct {
	Manager *ManagerService

	log       *log.Logger
	config    *StubConfig
	mu        sync.Mutex
	connected bool
	dials     []*grpc.ClientConn
}

func NewRuntimeStub(config *StubConfig) *RuntimeStub {
	return &RuntimeStub{
		log:       config.logger,
		config:    config,
		mu:        sync.Mutex{},
		connected: false,
		dials:     make([]*grpc.ClientConn, 0),
	}
}

func (n *RuntimeStub) Connect() error {
	n.mu.Lock()
	defer n.mu.Unlock()

	if n.connected {
		return nil
	}

	if n.config.manager.enabled {
		conn, services, err := NewManagerConnection(n.config.manager.url)
		if err != nil {
			n.log.Error("Error while connecting to the runtime_manager service: " + err.Error())
			n.closeConnections()
			return err
		}
		n.log.Info("Successfully connected to the runtime_manager service")
		n.dials = append(n.dials, conn)
		n.Manager = services
	}

	n.connected = true
	return nil
}

func (n *RuntimeStub) Close() {
	n.mu.Lock()
	defer n.mu.Unlock()
	n.closeConnections()
}

func (n *RuntimeStub) closeConnections() {
	for _, dial := range n.dials {
		dial.Close()
	}
	n.dials = make([]*grpc.ClientConn, 0)
	n.connected = false
}

File: modules/runtime/libs/ts/src/stub.ts
import { Client, dummyClient, Rpc } from './client'
import { Config, makeDefaultConfig } from './config'

import { RuntimeServiceClientImpl } from './proto/manager/runtime'
import { RPCServiceClientImpl } from './proto/manager/rpc'

interface RuntimeServices {
    manager: {
        runtime: RuntimeServiceClientImpl,
        rpc: RPCServiceClientImpl
    }
}

export enum Service {MANAGER}

export class RuntimeStub {
    private clients: Array<Client>
    public services: RuntimeServices

    constructor(connectTo: Array<Service> = [Service.MANAGER], config?: Config) {
        if (config === undefined) {
            config = makeDefaultConfig()
        }
        this.clients = []
        const clientByUrl = new Map<string, Client>()
        
        const connectToSet = new Set(connectTo)
        
        const grpcConfig = {
            'grpc.service_config': JSON.stringify({ loadBalancingConfig: [{ round_robin: {} }], })
        }
        const makeRpc = (service: Service, url: string) => {
            if (connectToSet.has(service)) {
                const savedCliet = clientByUrl.get(url)
                if (savedCliet !== undefined) return savedCliet
                
                const client = new Client(url, undefined, grpcConfig)
                this.clients.push(client)
                clientByUrl.set(url, client)
                return client
            }
            return dummyClient
        }
        
        const managerConnection = makeRpc(Service.MANAGER, config.urls.manager)

        this.services = {
            manager: {
                runtime: new RuntimeServiceClientImpl(managerConnection),
                rpc: new RPCServiceClientImpl(managerConnection)
            }
        }
    }

    async connect() {
        for(const client of this.clients) {
            await client.connect()
        }
    }

    close() {
        for(const client of this.clients) {
            client.close()
        }
    }
}